File: android/app/src/main/java/ai/oriveo/community/core/usage/CostSummaryCalculator.kt
package ai.oriveo.community.core.usage

import ai.oriveo.community.core.model.ChatMessageState
import ai.oriveo.community.core.model.ChatRole
import ai.oriveo.community.core.model.Conversation
import ai.oriveo.community.core.model.CostFormatter
import ai.oriveo.community.core.model.Provider
import ai.oriveo.community.core.model.ProviderKind
import ai.oriveo.community.core.model.resolveProviderLogoKind
import java.time.Instant
import java.time.ZoneOffset
import java.time.temporal.ChronoUnit

/** Where a spend figure came from. Only one source exists: this device's own message table. */
enum class CostSummarySource {
    LocalDevice,
}

/** This month's spend, split across the providers that account for most of it. */
data class MonthlyCostSummary(
    val totalCost: Double = 0.0,
    val providers: List<MonthlyCostProviderEntry> = emptyList(),
    val hiddenProviderCount: Int = 0,
    val source: CostSummarySource = CostSummarySource.LocalDevice,
) {
    val isVisible: Boolean
        get() = totalCost > CostFormatter.COST_EPSILON && providers.isNotEmpty()
}

data class MonthlyCostProviderEntry(
    val providerKind: ProviderKind,
    /** Kept alongside the kind so two relay connections do not merge into one row. */
    val providerID: String,
    /** The connection's own name where it has one, otherwise the vendor's. */
    val displayName: String,
    val logoProviderKind: ProviderKind? = null,
    val cost: Double,
)

object CostSummaryCalculator {
    private val entryOrder: Comparator<MonthlyCostProviderEntry> =
        compareByDescending<MonthlyCostProviderEntry> { it.cost }
            .thenBy { it.displayName }

    fun calculateMonthlySummary(
        conversations: List<Conversation>,
        providers: List<Provider>,
        now: Instant = Instant.now(),
        visibleProviderLimit: Int = 3,
    ): MonthlyCostSummary {
        val windowStart = now.atZone(ZoneOffset.UTC)
            .withDayOfMonth(1)
            .truncatedTo(ChronoUnit.DAYS)
            .toInstant()
        val windowEnd = windowStart.atZone(ZoneOffset.UTC)
            .plusMonths(1)
            .toInstant()

        val providerByID = providers.associateBy { it.id }

        val groupedCosts = mutableMapOf<String, Triple<ProviderKind, String, Double>>()

        conversations
            .asSequence()
            .filterNot { it.isDraft }
            .forEach { conversation ->
                conversation.messages.forEach { message ->
                    if (message.role != ChatRole.Assistant) return@forEach
                    if (message.state != ChatMessageState.Delivered) return@forEach
                    if (message.estimatedCost <= CostFormatter.COST_EPSILON) return@forEach

                    val occurredAt = message.createdAt ?: conversation.updatedAt
                    val occurredAtInstant = Instant.ofEpochMilli(occurredAt)
                    if (occurredAtInstant < windowStart || occurredAtInstant >= windowEnd) return@forEach

                    val providerID = message.providerID ?: conversation.providerID
                    val key = "${message.providerKind.name}|$providerID"
                    val previous = groupedCosts[key]
                    groupedCosts[key] = Triple(
                        message.providerKind,
                        providerID,
                        (previous?.third ?: 0.0) + message.estimatedCost,
                    )
                }
            }

        val sortedProviders = groupedCosts.values
            .map { (kind, providerID, cost) ->
                MonthlyCostProviderEntry(
                    providerKind = kind,
                    providerID = providerID,
                    displayName = providerByID[providerID]?.displayName ?: kind.displayName,
                    logoProviderKind = providerByID[providerID]?.let(::resolveProviderLogoKind),
                    cost = cost,
                )
            }
            .sortedWith(entryOrder)

        return MonthlyCostSummary(
            totalCost = sortedProviders.sumOf { it.cost },
            providers = sortedProviders.take(visibleProviderLimit),
            hiddenProviderCount = maxOf(0, sortedProviders.size - visibleProviderLimit),
            source = CostSummarySource.LocalDevice,
        )
    }

    /** Fills in each entry's display name and logo from the current provider list. */
    fun augmentDisplayNames(
        summary: MonthlyCostSummary,
        providers: List<Provider>,
    ): MonthlyCostSummary {
        if (providers.isEmpty()) return summary
        val byID = providers.associateBy { it.id }
        return summary.copy(
            providers = summary.providers.map { entry ->
                val provider = byID[entry.providerID]
                val displayName = provider?.displayName ?: entry.displayName
                val logoProviderKind = provider?.let(::resolveProviderLogoKind) ?: entry.logoProviderKind
                if (displayName == entry.displayName && logoProviderKind == entry.logoProviderKind) {
                    entry
                } else {
                    entry.copy(displayName = displayName, logoProviderKind = logoProviderKind)
                }
            },
        )
    }

    /** This month's spend keyed by provider id, for the provider list rows. */
    fun monthlyCostByProvider(
        conversations: List<Conversation>,
        now: Instant = Instant.now(),
    ): Map<String, Double> {
        val windowStart = now.atZone(ZoneOffset.UTC)
            .withDayOfMonth(1)
            .truncatedTo(ChronoUnit.DAYS)
            .toInstant()
        val windowEnd = windowStart.atZone(ZoneOffset.UTC)
            .plusMonths(1)
            .toInstant()

        val costs = mutableMapOf<String, Double>()

        conversations
            .asSequence()
            .filterNot { it.isDraft }
            .forEach { conversation ->
                conversation.messages.forEach { message ->
                    if (message.role != ChatRole.Assistant) return@forEach
                    if (message.state != ChatMessageState.Delivered) return@forEach
                    if (message.estimatedCost <= CostFormatter.COST_EPSILON) return@forEach

                    val occurredAt = message.createdAt ?: conversation.updatedAt
                    val occurredAtInstant = Instant.ofEpochMilli(occurredAt)
                    if (occurredAtInstant < windowStart || occurredAtInstant >= windowEnd) return@forEach

                    costs[conversation.providerID] =
                        (costs[conversation.providerID] ?: 0.0) + message.estimatedCost
                }
            }

        return costs
    }
}

File: ios/Oriveo/Oriveo/Core/Usage/CostSummaryCalculator.swift
import Foundation

struct MonthlyCostSummary: Equatable {
    var totalCost: Double = 0
    var providers: [MonthlyCostSummaryProviderEntry] = []
    var hiddenProviderCount: Int = 0

    var isVisible: Bool {
        totalCost > CostFormatter.costEpsilon && !providers.isEmpty
    }
}

struct MonthlyCostSummaryProviderEntry: Identifiable, Equatable {
    let providerKind: ProviderKind
    let providerID: String
    let displayName: String
    let logoProviderKind: ProviderKind?
    let cost: Double

    var id: String { "\(providerKind.rawValue)|\(providerID)" }
}

struct MonthlyCostAggregate: Equatable, Sendable {
    let providerKind: ProviderKind
    let providerID: UUID
    let cost: Double
}

enum CostSummaryCalculator {
    static func utcMonthBounds(containing date: Date) -> (start: TimeInterval, end: TimeInterval) {
        let window = utcMonthWindow(containing: date)
        return (window.start.timeIntervalSince1970, window.end.timeIntervalSince1970)
    }

    static func makeMonthlySummary(
        from conversations: [Conversation],
        providers: [Provider],
        now: Date = Date(),
        visibleProviderLimit: Int = 3
    ) -> MonthlyCostSummary {
        makeMonthlySummary(
            from: aggregates(from: conversations, now: now),
            providers: providers,
            visibleProviderLimit: visibleProviderLimit
        )
    }

    static func makeMonthlySummary(
        from aggregates: [MonthlyCostAggregate],
        providers: [Provider],
        visibleProviderLimit: Int = 3
    ) -> MonthlyCostSummary {
        let providerByID = Dictionary(uniqueKeysWithValues: providers.map { ($0.id, $0) })
        var groupedCosts: [String: MonthlyCostAggregate] = [:]
        for aggregate in aggregates where aggregate.cost > CostFormatter.costEpsilon {
            let key = "\(aggregate.providerKind.rawValue)|\(aggregate.providerID.uuidString)"
            if let previous = groupedCosts[key] {
                groupedCosts[key] = MonthlyCostAggregate(
                    providerKind: previous.providerKind,
                    providerID: previous.providerID,
                    cost: previous.cost + aggregate.cost
                )
            } else {
                groupedCosts[key] = aggregate
            }
        }

        let sortedProviders = groupedCosts.values
            .map { entry in
                MonthlyCostSummaryProviderEntry(
                    providerKind: entry.providerKind,
                    providerID: entry.providerID.uuidString,
                    displayName: providerByID[entry.providerID]?.displayName ?? entry.providerKind.displayName,
                    logoProviderKind: providerByID[entry.providerID].map(ProviderLogoResolver.logoKind(for:)),
                    cost: entry.cost
                )
            }
            .sorted(by: monthlyEntryOrder)

        return MonthlyCostSummary(
            totalCost: sortedProviders.map(\.cost).reduce(0, +),
            providers: Array(sortedProviders.prefix(visibleProviderLimit)),
            hiddenProviderCount: max(0, sortedProviders.count - visibleProviderLimit)
        )
    }

    static func aggregates(from conversations: [Conversation], now: Date = Date()) -> [MonthlyCostAggregate] {
        let window = utcMonthWindow(containing: now)
        var groupedCosts: [String: MonthlyCostAggregate] = [:]

        for conversation in conversations where !conversation.isDraft {
            for message in conversation.messages {
                guard message.role == .assistant,
                      message.state == .delivered,
                      message.estimatedCost > CostFormatter.costEpsilon
                else { continue }

                let occurredAt = message.createdAt ?? conversation.updatedAt
                guard isWithinWindow(occurredAt, window: window) else { continue }

                let providerID = message.providerID ?? conversation.providerID
                let key = "\(message.providerKind.rawValue)|\(providerID.uuidString)"
                if let previous = groupedCosts[key] {
                    groupedCosts[key] = MonthlyCostAggregate(
                        providerKind: previous.providerKind,
                        providerID: previous.providerID,
                        cost: previous.cost + message.estimatedCost
                    )
                } else {
                    groupedCosts[key] = MonthlyCostAggregate(
                        providerKind: message.providerKind,
                        providerID: providerID,
                        cost: message.estimatedCost
                    )
                }
            }
        }
        return Array(groupedCosts.values)
    }



    private static func monthlyEntryOrder(
        _ lhs: MonthlyCostSummaryProviderEntry,
        _ rhs: MonthlyCostSummaryProviderEntry
    ) -> Bool {
        if lhs.cost != rhs.cost { return lhs.cost > rhs.cost }
        return lhs.displayName.localizedStandardCompare(rhs.displayName) == .orderedAscending
    }

    static func monthlyCostByProvider(
        from conversations: [Conversation],
        now: Date = Date()
    ) -> [UUID: Double] {
        monthlyCostByProvider(from: aggregates(from: conversations, now: now))
    }

    static func monthlyCostByProvider(from aggregates: [MonthlyCostAggregate]) -> [UUID: Double] {
        var costs: [UUID: Double] = [:]
        for aggregate in aggregates where aggregate.cost > CostFormatter.costEpsilon {
            costs[aggregate.providerID, default: 0] += aggregate.cost
        }
        return costs
    }

    private static func utcMonthWindow(containing date: Date) -> DateInterval {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0) ?? .gmt

        let components = calendar.dateComponents([.year, .month], from: date)
        let start = calendar.date(from: components) ?? date
        let end = calendar.date(byAdding: .month, value: 1, to: start) ?? date
        return DateInterval(start: start, end: end)
    }

    private static func isWithinWindow(_ date: Date, window: DateInterval) -> Bool {
        date >= window.start && date < window.end
    }
}
